src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -191,37 +193,97 @@ public bool PrepareForReuse(bool async)
+                bool restoreFlow = false;
+                try
+                {
+                    if (!ExecutionContext.IsFlowSuppressed())
+                    {
+                        ExecutionContext.SuppressFlow();
+                        restoreFlow = true;
+                    }
 
-            // We may already have a read-ahead task if we did a previous scavenge and haven't used the connection since.
 #pragma warning disable CA2012 // we're very careful to ensure the ValueTask is only consumed once, even though it's stored into a field
-            _readAheadTask ??= ReadAheadWithZeroByteReadAsync();
+                    _readAheadTask = ReadAheadWithZeroByteReadAsync();
 #pragma warning restore CA2012
-
-            // If the read-ahead task is completed, then we've received either EOF or erroneous data the connection, so it's not usable.
-            return !_readAheadTask.Value.IsCompleted;
+                }
+                finally
+                {
+                    if (restoreFlow)
+                    {
+                        ExecutionContext.RestoreFlow();
+                    }
+                }
+            }
 
             async ValueTask<int> ReadAheadWithZeroByteReadAsync()
             {
-                Debug.Assert(_readAheadTask is null);
+                Debug.Assert(_readAheadTask == default);
                 Debug.Assert(_readBuffer.ActiveLength == 0);
 
-                // Issue a zero-byte read.
-                // If the underlying stream supports it, this will not complete until the stream has data available,
-                // which will avoid pinning the connection's read buffer (and possibly allow us to release it to the buffer pool in the future, if desired).
-                // If not, it will complete immediately.
-                await _stream.ReadAsync(Memory<byte>.Empty).ConfigureAwait(false);
+                try
+                {
+                    // Issue a zero-byte read.
+                    // If the underlying stream supports it, this will not complete until the stream has data available,
+                    // which will avoid pinning the connection's read buffer (and possibly allow us to release it to the buffer pool in the future, if desired).
+                    // If not, it will complete immediately.
+                    await _stream.ReadAsync(Memory<byte>.Empty).ConfigureAwait(false);
+
+                    // We don't know for sure that the stream actually has data available, so we need to issue a real read now.
+                    int read = await _stream.ReadAsync(_readBuffer.AvailableMemory).ConfigureAwait(false);
+
+                    if (TryOwnReadAheadTaskCompletion())
+                    {
+                        if (NetEventSource.Log.IsEnabled()) Trace("Read-ahead task observed data before the request was sent.");
+
+                        Dispose(disposing: true, fromReadAheadTask: true);
```

Comment:
Why do we want to do this, rather than just letting 0 be returned to read in SendAsync?

Reply:
Related: https://github.com/dotnet/runtime/pull/80214#discussion_r1088013175 -- to handle the connection cleanup immediately instead of waiting for either the next request or scavenging logic.